src/array.jl

```diff
@@ -972,3 +972,31 @@ Base.repeat(a::CatArrOrSub{T, N},
 Base.repeat(a::CatArrOrSub{T, N};
             inner = nothing, outer = nothing) where {T, N} =
     CategoricalArray{T, N}(repeat(refs(a), inner=inner, outer=outer), copy(pool(a)))
+
+# JSON3 writing/reading
+StructTypes.StructType(::Type{<:CategoricalVector}) = StructTypes.ArrayType()
+
+StructTypes.construct(::Type{<:CategoricalVector}, array::Vector) =
```

Comment:
This method is redundant as it's a special case of the next one AFAICT. Same below.

Reply:
I thought so too; However, without this definition I run into problems. For instance:

```
julia>  x = CategoricalArray([missing,"y","z","y",missing,"z","x"])
julia> str = JSON3.write(x)
julia> readx = JSON3.read(str, CategoricalVector)
7-element CategoricalArray{Any,1,UInt32}:
 nothing
 "y"
 "z"
 "y"
 nothing
 "z"
 "x"
```

The result returns `nothing` instead of `missing` because `categoricalgeneral` never gets called, and a generic fallback in `JSON3` is called instead.

Reply:
That's weird. Can you see what arguments are passed to `construct` in that case? Also, how does `JSON3.read` manage to return a `CategoricalArray` if it doesn't call the method defined here?

Reply:
Here is the default constructor and the values passed to it in this case:

```julia
construct(T, x::Vector{S}; kw...) where {S} = T(x)
T = CategoricalArray{T,1,V,C,U,U1} where {U1, U, C, V, T}
x = Any[nothing, "y", "z", "y", nothing, "z", "x"]
```

Reply:
Indeed:

```julia
julia> (CategoricalArray{T,1,V,C,U,U1} where U1 where U where C where V where T) <: CategoricalArray
false
```

:sob:

Reply:
Looks like you need `R<:Integer`:
```julia
julia> (CategoricalArray{T,1,R,C,V,U} where T where R<:Integer where V where C where U) <: CategoricalArray
true
```

Reply:
Ah okay, I see; because of the constraints defined for [`CategoricalArray`](https://github.com/JuliaData/CategoricalArrays.jl/blob/master/src/typedefs.jl#L103), `CategoricalArray` includes this constraint, but `CategoricalVector`, on its own, does not do this.

Do you think there is a better solution to this then just repeating the method as I've done here?

Reply:
Hmm, looks like there's just a bug in the definition of `CategoricalVector`. It should be
```julia
CategoricalVector{T, R, V, C, U} = CategoricalArray{T, 1, R, V, C, U}
```
or maybe (if necessary to get dispatch to work):
```julia
CategoricalVector{T, R, V, C, U} = CategoricalArray{T, 1, R <: Integer, V, C, U}
```

Same for `CategoricalMatrix`.

Reply:
Hmm... changing to the first does not change dispatch, and changing to the second leads to a `StackOverflowError` when constructing a categorical array at line array.jl:258. I can look into it later today or tomorrow.